### PR TITLE
feat: Create useDismissAlert help react hook to replace some jank

### PR DESCRIPTION
**Pull request by ryan953**
It was a manual process to deal with dismissing the React 'inspect your dom' alert.

We can do better by putting that behavior into a hook, and it'll be re-usable for some replay update/onboarding stuff coming up.

### CODE REVIEW

File: static/app/utils/useDismissAlert.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  /**
+   * Number of days before the dismissal is expired.
+   * After expiration the the user will need to re-dismiss things.
+   */
+  expirationDays?: number;
+};
+
+const MS_PER_DAY = 1000 * 60 * 60 * 24;
+
+function useDismissAlert({expirationDays = Number.MAX_SAFE_INTEGER, key}: Opts) {
```
Comment (evanpurkhiser) [inline static/app/utils/useDismissAlert.tsx:21]: Could we add some description to this? It sounds pretty generic

Comment (ryan953) [reply static/app/utils/useDismissAlert.tsx:30]: added!

File: static/app/utils/useDismissAlert.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * Dismissal can be 'permanent' or have an expiration after some number of days.
+ *
+ * The real lifecycle depends on whether users reset localStorage or not, if you
+ * need something really permanent then save the users' preference to the server.
+ */
+function useDismissAlert({expirationDays = Number.MAX_SAFE_INTEGER, key}: Opts) {
+  const [isDismissed, setDismissed] = useState(false);
+
+  useEffect(() => {
+    const val = localStorage.getItem(key);
```
Comment (evanpurkhiser) [inline static/app/utils/useDismissAlert.tsx:34]: Would it make sense to use the `useLocalStorageState` hook?

File: static/app/utils/useDismissAlert.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    expect(mockSetItem).toHaveBeenCalledWith(key, String(now.getTime()));
+  });
+
+  it('should be dismissed if the timestamp in localstorage is older than the expiration', () => {
+    const today = new Date('2020-01-01');
+    jest.setSystemTime(today);
+
+    // Dismissed on christmas
+    const christmas = new Date('2019-12-25').getTime();
```
Comment (evanpurkhiser) [inline static/app/utils/useDismissAlert.spec.tsx:80]: Why not `toString()`?

File: static/app/utils/useDismissAlert.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
+import TestRenderer from 'react-test-renderer';
```
Comment (evanpurkhiser) [inline static/app/utils/useDismissAlert.spec.tsx:1]: What is this?

File: static/app/utils/useDismissAlert.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      const timestamp = Number(val);
+      if (isNaN(timestamp)) {
+        localStorage.setItem(key, String(Date.now()));
+        setDismissed(true);
+      } else {
+        const duration = Date.now() - timestamp;
+        const days = duration / MS_PER_DAY;
+        setDismissed(days < expirationDays);
+      }
+    }
```
Comment (evanpurkhiser) [inline static/app/utils/useDismissAlert.tsx:48]: kinda a nit but I tend to find this easier to read

```suggestion
    if (expirationDays === Number.MAX_SAFE_INTEGER) {
      setDismissed(!!val);
      return;
    }
    
    if (!val) {
    	return
    }
    
    const timestamp = Number(val);
    if (isNaN(timestamp)) {
      localStorage.setItem(key, Date.now().toString());
      setDismissed(true);
    } else {
      const duration = Date.now() - timestamp;
      const days = duration / MS_PER_DAY;
      setDismissed(days < expirationDays);
    }
```

File: static/app/utils/useDismissAlert.tsx (hunk 2 above)
Comment (ryan953) [reply static/app/utils/useDismissAlert.tsx:34]: yeah, that's a handy hook. 
It's frustrating to test that something was actually saved to localStorage though, I don't see any tests that wire things up to assert that kind of thing either. Something about the queueMicrotask call, the `jest.run*` methods don't seem to help either :(

Comment (evanpurkhiser) [reply static/app/utils/useDismissAlert.tsx:34]: you could rely on it's tests right and not need to test that it's actually putting stuff in local storage, instead you can spy on it's functions and check that those are called maybe?

File: static/app/utils/useDismissAlert.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import {useCallback} from 'react';
+
+import {useLocalStorageState} from 'sentry/utils/useLocalStorageState';
+
+type Opts = {
+  /**
+   * Key in localstorage.
```
Comment (priscilawebdev) [inline static/app/utils/useDismissAlert.tsx:7]:
```suggestion
   * Key in localStorage.
```

File: static/app/utils/useDismissAlert.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    return false;
+  }
+  const duration = Date.now() - timestamp;
+  const days = duration / MS_PER_DAY;
+  return days < expirationDays;
+}
+
+/**
+ * Nominally for tracking dismissal/acknowledgement of in-app alerts and
+ * notifications, specfically those that don't need to be persisted server side.
```
Comment (priscilawebdev) [inline static/app/utils/useDismissAlert.tsx:32]:
```suggestion
 * notifications, specifically those that don't need to be persisted server side.
```

File: static/app/utils/useDismissAlert.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    setDismissedTimestamp(Date.now().toString());
+  }, [setDismissedTimestamp]);
+
+  return {
+    isDismissed,
+    dismiss,
+  };
+}
+
+export default useDismissAlert;
```
Comment (priscilawebdev) [inline static/app/utils/useDismissAlert.tsx:60]: nit: shall we make this `named export`?

File: static/app/utils/useDismissAlert.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+      jest.runAllTicks();
+    });
+
+    expect(mockSetItem).toHaveBeenCalledWith(
+      key,
+      JSON.stringify(now.getTime().toString())
+    );
+  });
+
+  it('should be dismissed if the timestamp in localstorage is older than the expiration', () => {
```
Comment (priscilawebdev) [inline static/app/utils/useDismissAlert.spec.tsx:76]:
```suggestion
  it('should be dismissed if the timestamp in localStorage is older than the expiration', () => {
```

File: static/app/utils/useDismissAlert.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    const today = new Date('2020-01-01');
```
Comment (priscilawebdev) [inline static/app/utils/useDismissAlert.spec.tsx:77]: should we reuse the const `now` here?

File: static/app/components/replays/playerDOMAlert.tsx [hunk 11]
```diff
@@ -1,26 +1,15 @@
...
-  const handleDismiss = () => {
-    localStorage.setItem(LOCAL_STORAGE_KEY, '1');
-    setIsDismissed(true);
-  };
-
-  useEffect(() => {
-    const val = localStorage.getItem(LOCAL_STORAGE_KEY);
-    setIsDismissed(val === '1');
-  }, []);
+  const {dismiss, isDismissed} = useDismissAlert({key: LOCAL_STORAGE_KEY});
```
Comment (priscilawebdev) [inline static/app/components/replays/playerDOMAlert.tsx:12]: hm, shouldn't this be named `onDismiss` to indicate that an event?

Comment (priscilawebdev) [reply static/app/components/replays/playerDOMAlert.tsx:12]: never mind... I just noticed that React Query names event handlers like this too  🤷‍♀️ https://tanstack.com/query/v4/docs/react/reference/useQuery

File: static/app/utils/useDismissAlert.tsx (hunk 5 above)
Comment (priscilawebdev) [reply static/app/utils/useDismissAlert.tsx]: I agree with Evan :) I personally prefer using `early return` than `else` when possible

File: static/app/utils/useDismissAlert.spec.tsx (hunk 10 above)
Comment (ryan953) [reply static/app/utils/useDismissAlert.spec.tsx:77]: i used a new variable so it's local to the test and easier to spot the difference between the two dates.

File: static/app/utils/useDismissAlert.tsx (hunk 8 above)
Comment (ryan953) [reply static/app/utils/useDismissAlert.tsx:60]: If that's not a hard rule i'm not really interested in doing it. I have opinions on named exports but this isn't the place for that

**Review by markstory — COMMENTED** [review]
Looks good to me, would be good to replace some of the other dismissable UI alerts with this hook as I bet they all have bespoke implementations right now.

File: static/app/utils/useDismissAlert.tsx (hunk 8 above)
Comment (priscilawebdev) [reply static/app/utils/useDismissAlert.tsx:60]: agree...maybe that is something for the Frontend TSC

**Review by priscilawebdev — COMMENTED** [review]
Looks good to me. Are you planning to replace other dismissable UI alerts in a follow-up?

**Review by priscilawebdev — APPROVED** [review]

**Comment by ryan953** [comment]
> Looks good to me. Are you planning to replace other dismissable UI alerts in a follow-up?

I looked at some already to replace. I found one snoozable instance that's a little tricky because the data it saves is different from the timestamps that this saves, so we'd have to read back the old val and update what we find.

**Pull request merged**